packages/react-components/react-dialog/Spec.md

```diff
@@ -1,63 +1,524 @@
+/>
+```
 
 ## API
 
-_List the **Props** and **Slots** proposed for the component. Ideally this would just be a link to the component's `.types.ts` file_
+The `Dialog` should implement a `children` based API as is one of the standards across all the surveyed alternatives as a part of Open UI research in [Prior Art](#prior-art). The component will leverage the use of `context` in the interaction and data flows of child compound components.
 
-## Structure
+Sample usages will be give in the following section of this document [Sample code](#sample-code)
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+### Dialog
 
-## Migration
+The root level component serves as an interface for interaction with all possible behaviors exposed. It provides context down the hierarchy to `children` compound components to allow functionality. This component expects to receive as children either a `DialogContent` or a `DialogTrigger` and a `DialogContent` (or some component that will eventually render one of those compound components) in this specific order
+
+```tsx
+type DialogSlots = {
+  /**
+   * The dialog element itself
+   */
+  root: Slot<'div'>;
+  /**
+   * Dimmed background of dialog.
+   * The default overlay is rendered as a `<div>` with styling.
+   * This slot expects a `<div>` element which will replace the default overlay.
+   * The overlay should have `aria-hidden="true"`.
+   */
+  overlay?: Slot<'div'>;
+  /**
+   * The root element of the Dialog right after Portal.
+   */
+  root: Slot<'div'>;
+};
+
+type DialogProps = ComponentProps<DialogSlots> & {
+  /**
+   * Dialog variations.
+   *
+   * `modal`: When this type of dialog is open, the rest of the page is dimmed out and cannot be interacted with. The tab sequence is kept within the dialog and moving the focus outside the dialog will imply closing it. This is the default type of the component.
+   *
+   * `non-modal`: When a non-modal dialog is open, the rest of the page is not dimmed out and users can interact with the rest of the page. This also implies that the tab focus can move outside the dialog when it reaches the last focusable element.
+   *
+   * `alert`: is a special type of modal dialogs that interrupts the user's workflow to communicate an important message or ask for a decision. Unlike a typical modal dialog, the user must take an action through the options given to dismiss the dialog, and it cannot be dismissed through the dimmed background or escape key.
+   */
+  type?: 'modal' | 'non-modal' | 'alert';
+  /**
+   * Controls the open state of the dialog
+   * @default undefined
+   */
+  open?: boolean;
+  /**
+   * Default value for the uncontrolled open state of the dialog.
+   * @default false
+   */
+  defaultOpen?: boolean;
+  /**
+   * Callback fired when the component changes value from open state.
+   * @default undefined
+   */
+  onOpenChange?(event: MouseEvent | KeyboardEvent, data: DialogOpenChangeData): void;
```

Comment:
It could be React and native events there, correct?

Reply:
Unfortunately, yes. In the rare use case where the Dialog is defined with absolutely no focusable element inside itself, a listener on the document might be used to ensure `Escape` functionality properly, and in that case a normal event will be triggered instead of a React one, but this is IMO more of an implementation detail than a Spec thing.